process_hierarchy.py

```diff
@@ -0,0 +1,451 @@
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+            )
+            hierarchy_file.write(f"{unique_key},{l3_rec['name']},{parent_key}\n")
+
+    # constrcut groups.csv (level 4 of the hierarchy, though this is a mapping table between pi/group and the hiearachy)
+    with open("group.csv", "w", encoding="utf-8") as group_file:
+        for l4_id, l4_rec in hierarchy["project"].items():
+            l3_rec = hierarchy["pi"][l4_rec["parent_id"]]
+            unique_key = create_xdmod_key(
+                (hierarchy["field-of-science"][l3_rec["parent_id"]])["parent_id"],
+                l3_rec["parent_id"],
+                l3_rec["id"],
+            )
+            group_file.write(f'"{l4_rec["name"]}", "{unique_key}" \n')
+
+    # constrcut pi2project
+    with open("pi2project.csv", "w", encoding="utf-8") as pi2project_file:
+        for l5 in hierarchy["cloud-project"].values():
+            l4_rec = hierarchy["cloud-project"][l5["parent_id"]]
+            l4_id = l4_rec["id"]
+            pi2project_file.write(f"{hierarchy['cloud-project'][l4_id]['name']}, {l5['name']}\n")
+
+    # construct the names (rename projects)
+
+
+def find_hierarchy_id(name, dictionary):
+    """returns the hierarchy_id if found in the dictionary"""
+    for rec in dictionary.values():
+        if name == rec["name"]:
+            return rec["id"]
+    return None
+
+
+def get_hierarchy_from_db(cursor):
+    """This fetches the data from the database and only keeps the most recent record based on its id"""
+    hierarchy = {}
+    hierarchy["institution"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='institution' order by hr.create_ts",
+        None,
+    )
+    hierarchy["field-of-science"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='field-of-science' order by hr.create_ts",
+        None,
+    )
+    hierarchy["pi"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='pi' order by hr.create_ts",
+        None,
+    )
+    hierarchy["project"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='openstack-project' or hr.type='openshift-project' order by hr.create_ts",
+        None,
+    )
+    hierarchy["cloud-project"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='openstack-project' order by hr.create_ts",
+        None,
+    )
+    return hierarchy
+
+
+def process_expected_data(cursor, hierarchy):
+    """This fits idealized data into the hierarchy from the database
+
+    This is currently needed as we are not getting the data from
+    keycloak yet.
+    """
+    data = get_data()
+    for record in data:
+        match record["type"]:
```

Comment:
Switch statements are a new feature in Python 3.10, and this image uses Python 3.8, so this is going to throw a syntax error.